src/topology/algebra/order/lower.lean

```diff
@@ -0,0 +1,271 @@
+
+/--
+Type synonym for a preorder equipped with the lower topology
+-/
+@[reducible]
+def lower (α : Type u) := α
+
+-- Needed if the above definition is not reducible
+/-
+instance lower.preorder (α : Type u) [preorder α] :
+  preorder (lower α) := _inst_1
+-/
+
+/--
+The lower topology is the topology generated by the complements of the closed intervals to infinity
+-/
+def lower_topology (α : Type u) [preorder α] : topological_space α := generate_from {s | ∃ a, (Ici a)ᶜ = s}
+
+instance lower.topological_space (α : Type u) [preorder α] :
+  topological_space (lower α) := lower_topology α
+
+
+section pre_order
+
+
+variable [preorder α]
+
+lemma lower_topology_def : (lower_topology α) = generate_from {s | ∃ a, (Ici a)ᶜ = s} := rfl
+
+lemma is_open_iff_generate_Ici_comp {s : set α} :
+  is_open s ↔ generate_open {s | ∃a, (Ici a)ᶜ = s } s := iff.rfl
+
+/-
+Left-closed right-infinite intervals [a,∞) are closed in the lower topology
+-/
+lemma ici_is_closed (a : α) :
+  is_closed (Ici a)  :=
+begin
+  rw [← is_open_compl_iff, is_open_iff_generate_Ici_comp],
+  fconstructor,
+  rw mem_set_of_eq,
+  use a,
+end
+
+/-
+The upper closure of a finite subset is closed in the lower topology
+-/
+lemma upper_closure_is_closed (F : set α) (h : F.finite) : is_closed (upper_closure F : set α) :=
+begin
+  rw ← upper_set.infi_Ici,
+  simp only [upper_set.coe_infi, upper_set.coe_Ici],
+  apply is_closed_bUnion h,
+  intros a h₁,
+  apply ici_is_closed,
+end
+
+/-
+Every subset open in the lower topology is a lower set
+-/
+lemma lower_open_is_lower {s : set α} (h: is_open s) : is_lower_set s :=
+begin
+  rw is_open_iff_generate_Ici_comp at h,
+  induction h,
+  case topological_space.generate_open.basic : u
+  { rw mem_set_of_eq at h_H,
+    choose a h_H using h_H,
+    rw ← h_H,
+    apply is_upper_set.compl,
+    apply is_upper_set_Ici, },
+  case topological_space.generate_open.univ : { exact is_lower_set_univ },
+  case topological_space.generate_open.inter : u v hu1 hv1 hu2 hv2
+    { apply is_lower_set.inter hu2 hv2 },
+  case topological_space.generate_open.sUnion : { apply is_lower_set_sUnion h_ih, },
+end
+
+/-
+The closure of a singleton {a} in the lower topology is the left-closed right-infinite interval
+[a,∞)
+-/
+lemma singleton_closure (a : α) : closure {a} = Ici a :=
+begin
+  rw subset_antisymm_iff,
+  split,
+  { apply closure_minimal _ (ici_is_closed a), rw [singleton_subset_iff, mem_Ici], },
+  { unfold closure,
+    refine subset_sInter _,
+    intro u,
+    intro h,
+    rw mem_set_of_eq at h,
+    intro b,
+    intro hb,
+    rw mem_Ici at hb,
+    rw [singleton_subset_iff, ← is_open_compl_iff] at h,
+    by_contradiction H,
+    rw ← mem_compl_iff at H,
+    have h1: a ∈ uᶜ, from lower_open_is_lower h.left hb H,
+    rw mem_compl_iff at h1,
+    rw ← not_not_mem at h,
+    apply absurd h1 h.right, },
+end
+
+end pre_order
+
+section partial_order
+
+variable [partial_order α]
+
+lemma Ici_eq (a b : α) : Ici a = Ici b ↔ a = b :=
```

Comment:
This lemma should be moved and called `set.Ici_inj` probably.

Reply:
Dove via #17157, thanks.